File: clients/go/client_test.go
package zedclient

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"path/filepath"
	"strings"
	"testing"
)

func newTestClient(t *testing.T, base string) *Client {
	t.Helper()
	client, err := New(base)
	if err != nil {
		t.Fatal(err)
	}
	return client
}

func TestPathsMatchContract(t *testing.T) {
	if got := PackagePath("acme", "kit"); got != "/v1/packages/acme/kit" {
		t.Fatalf("PackagePath = %q", got)
	}
	if got := VersionPath("acme", "kit", "1.2.0"); got != "/v1/packages/acme/kit/versions/1.2.0" {
		t.Fatalf("VersionPath = %q", got)
	}
	if got := YankPath("acme", "kit", "1.2.0"); got != "/v1/packages/acme/kit/versions/1.2.0/yank" {
		t.Fatalf("YankPath = %q", got)
	}
	if got := VersionPath("acme", "kit", "release candidate/1"); got != "/v1/packages/acme/kit/versions/release%20candidate%2F1" {
		t.Fatalf("escaped VersionPath = %q", got)
	}
	if got := PackagePath("a?b", "c#d"); got != "/v1/packages/a%3Fb/c%23d" {
		t.Fatalf("escaped PackagePath = %q", got)
	}
}

func TestRegistryURLIsValidatedAndTokenIsRedacted(t *testing.T) {
	client := newTestClient(t, " https://registry.zpkg.tech/gateway/// ")
	client.Token = "very-secret"
	if client.Base != "https://registry.zpkg.tech/gateway" {
		t.Fatalf("Base = %q", client.Base)
	}
	if strings.Contains(client.String(), "very-secret") || !strings.Contains(client.String(), "[REDACTED]") {
		t.Fatalf("unsafe client String: %s", client.String())
	}
	for _, invalid := range []string{
		"relative/path",
		"ftp://registry.test",
		"https://user:secret@registry.test",
		"https://registry.test?tenant=one",
		"https://registry.test#fragment",
	} {
		if _, err := New(invalid); err == nil {
			t.Fatalf("accepted invalid URL %q", invalid)
		}
	}
}

func TestPublicReadsOmitAuthAndAuthenticatedWritesAttachIt(t *testing.T) {
	var searchAuth, yankAuth, yankPathSeen, yankBody string
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/gateway/v1/search":
			searchAuth = r.Header.Get("Authorization")
			_, _ = w.Write([]byte(`{"query":"kit","items":[]}`))
		case "/gateway/v1/packages/acme/kit/versions/1.2.0/yank":
			yankAuth = r.Header.Get("Authorization")
			yankPathSeen = r.URL.EscapedPath()
			payload, _ := readBounded(r.Body, 1024, true)
			yankBody = string(payload)
			_, _ = w.Write([]byte(`{"org":"acme","name":"kit","version":"1.2.0","yanked":true}`))
		default:
			http.NotFound(w, r)
		}
	}))
	defer server.Close()

	client := newTestClient(t, server.URL+"/gateway///")
	client.Token = " zpkg_t "
	if _, err := client.Search("kit"); err != nil {
		t.Fatal(err)
	}
	result, err := client.Yank("acme", "kit", "1.2.0", true)
	if err != nil {
		t.Fatal(err)
	}
	if !result.Yanked || searchAuth != "" || yankAuth != "Bearer zpkg_t" {
		t.Fatalf("unexpected auth/result: search=%q yank=%q result=%+v", searchAuth, yankAuth, result)
	}
	if yankPathSeen != "/gateway/v1/packages/acme/kit/versions/1.2.0/yank" {
		t.Fatalf("yank path = %q", yankPathSeen)
	}
	var body map[string]bool
	if err := json.Unmarshal([]byte(yankBody), &body); err != nil || !body["yanked"] {
		t.Fatalf("yank body = %q, err=%v", yankBody, err)
	}
}

func TestRedirectsAreRefused(t *testing.T) {
	destination := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		t.Fatal("redirect destination must not be reached")
	}))
	defer destination.Close()
	source := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		http.Redirect(w, r, destination.URL, http.StatusFound)
	}))
	defer source.Close()

	client := newTestClient(t, source.URL)
	client.Token = "secret"
	_, err := client.ClaimOrg("acme")
	if !errors.Is(err, errRefuseRedirect) && !strings.Contains(err.Error(), "refuses redirects") {
		t.Fatalf("expected redirect refusal, got %v", err)
	}
}

func TestAPIErrorIsBoundedAndDefaultStringExcludesRemoteBody(t *testing.T) {
	remote := strings.Repeat("provider-secret", maxErrorBodyBytes)
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusBadGateway)
		_, _ = w.Write([]byte(remote))
	}))
	defer server.Close()
	client := newTestClient(t, server.URL)
	_, err := client.Search("x")
	var apiErr *APIError
	if !errors.As(err, &apiErr) {
		t.Fatalf("want APIError, got %T: %v", err, err)
	}
	if apiErr.Code != "http_502" || len(apiErr.Message) > maxErrorBodyBytes {
		t.Fatalf("unexpected bounded error: %+v", apiErr)
	}
	if strings.Contains(err.Error(), "provider-secret") {
		t.Fatal("default error string leaked remote body")
	}
}

func TestPackageMetadataVersionScheme(t *testing.T) {
	var meta PackageMetadata
	payload := `{"org":"acme","name":"kit","vcs":"git","repo_url":"x","versions":[],"version_scheme":"calver","tags":["http"]}`
	if err := json.Unmarshal([]byte(payload), &meta); err != nil {
		t.Fatal(err)
	}
	if meta.VersionScheme != "calver" || len(meta.Tags) != 1 {
		t.Fatalf("metadata = %+v", meta)
	}
}

func TestDownloadArtifactOmitsAuthResolvesRelativeAndVerifies(t *testing.T) {
	body := []byte("artifact-bytes")
	sum := sha256.Sum256(body)
	var gotAuth, gotPath string
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotAuth = r.Header.Get("Authorization")
		gotPath = r.URL.EscapedPath()
		_, _ = w.Write(body)
	}))
	defer server.Close()

	client := newTestClient(t, server.URL+"/gateway")
	client.Token = "zpkg_t"
	version := &VersionMetadata{
		Sha256:      hex.EncodeToString(sum[:]),
		Size:        uint64(len(body)),
		DownloadURL: "artifact/file",
	}
	if err := client.DownloadArtifact(version, filepath.Join(t.TempDir(), "artifact.tar.gz")); err != nil {
		t.Fatal(err)
	}
	if gotAuth != "" || gotPath != "/gateway/artifact/file" {
		t.Fatalf("download auth/path = %q / %q", gotAuth, gotPath)
	}
}

func TestDownloadArtifactStatusPolicyAndCaps(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(http.StatusNotFound)
		_, _ = w.Write([]byte(`{"code":"not_found","message":"no such artifact"}`))
	}))
	defer server.Close()
	client := newTestClient(t, server.URL)
	client.Token = "zpkg_t"
	err := client.DownloadArtifact(
		&VersionMetadata{Sha256: "abc", DownloadURL: server.URL + "/artifact"},
		filepath.Join(t.TempDir(), "artifact.tar.gz"),
	)
	var apiErr *APIError
	if !errors.As(err, &apiErr) || apiErr.Code != "not_found" || apiErr.Message != "no such artifact" {
		t.Fatalf("unexpected API error: %v / %+v", err, apiErr)
	}
	if strings.Contains(err.Error(), "no such artifact") {
		t.Fatal("default error leaked remote message")
	}

	// An explicitly insecure registry may use ordinary HTTP for its own artifact
	// URLs. Verify rejection from the production HTTPS trust boundary instead of
	// accidentally asking the network to resolve an intentionally bogus host.
	secureClient := newTestClient(t, "https://registry.zpkg.tech")
	for _, target := range []string{"http://evil.example/artifact", "file:///etc/passwd"} {
		err := secureClient.DownloadArtifact(
			&VersionMetadata{Sha256: "abc", DownloadURL: target},
			filepath.Join(t.TempDir(), "a"),
		)
		if !errors.As(err, &apiErr) || apiErr.Code != "insecure_download_url" {
			t.Fatalf("target %q: got %v", target, err)
		}
	}

	limit := downloadLimit(1)
	large := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_, _ = w.Write(make([]byte, limit+64))
	}))
	defer large.Close()
	largeClient := newTestClient(t, large.URL)
	err = largeClient.DownloadArtifact(
		&VersionMetadata{Sha256: "abc", Size: 1, DownloadURL: large.URL + "/artifact"},
		filepath.Join(t.TempDir(), "a"),
	)
	if !errors.As(err, &apiErr) || apiErr.Code != "artifact_too_large" {
		t.Fatalf("want artifact_too_large, got %v", err)
	}
}

func TestDownloadArtifactAllowsLoopbackHTTP(t *testing.T) {
	body := []byte("ok")
	sum := sha256.Sum256(body)
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		_, _ = w.Write(body)
	}))
	defer server.Close()
	client := newTestClient(t, "https://registry.zpkg.tech")
	version := &VersionMetadata{
		Sha256:      hex.EncodeToString(sum[:]),
		Size:        uint64(len(body)),
		DownloadURL: server.URL + "/artifact",
	}
	if err := client.DownloadArtifact(version, filepath.Join(t.TempDir(), "a")); err != nil {
		t.Fatalf("loopback HTTP download: %v", err)
	}
}

File: clients/typescript/test/client.test.mjs
import test from "node:test";
import assert from "node:assert/strict";
import { createHash } from "node:crypto";
import {
  ZedClient,
  ZedApiError,
  MAX_ARTIFACT_BYTES,
  MAX_ERROR_BODY_BYTES,
  packagePath,
  versionPath,
  yankPath,
  filePath,
  normalizeRegistryUrl,
} from "../dist/index.js";

function makeVersion(overrides = {}) {
  return {
    org: "acme",
    name: "kit",
    version: "1.2.0",
    sha256: "",
    size: 0,
    format: "tar.gz",
    vcs_tag: "v1.2.0",
    download_url: "",
    published_at: "2024-01-01T00:00:00Z",
    yanked: false,
    ...overrides,
  };
}

const sha256Hex = (bytes) => createHash("sha256").update(bytes).digest("hex");

test("url helpers match and encode the core contract", () => {
  assert.equal(packagePath("acme", "kit"), "/v1/packages/acme/kit");
  assert.equal(versionPath("acme", "kit", "1.2.0"), "/v1/packages/acme/kit/versions/1.2.0");
  assert.equal(yankPath("acme", "kit", "1.2.0"), "/v1/packages/acme/kit/versions/1.2.0/yank");
  assert.equal(filePath("acme", "kit", "1.2.0", "dist/x.css"), "/v1/files/acme/kit/1.2.0/dist/x.css");
  assert.equal(
    versionPath("acme", "kit", "release candidate/1"),
    "/v1/packages/acme/kit/versions/release%20candidate%2F1",
  );
  assert.equal(packagePath("a?b", "c#d"), "/v1/packages/a%3Fb/c%23d");
});

test("registry base URLs are validated and gateway prefixes are preserved", () => {
  assert.equal(normalizeRegistryUrl(" https://registry.test/gateway/// "), "https://registry.test/gateway");
  for (const invalid of [
    "relative/path",
    "ftp://registry.test",
    "https://user:secret@registry.test",
    "https://registry.test?tenant=one",
    "https://registry.test#fragment",
  ]) {
    assert.throws(() => new ZedClient({ registryUrl: invalid }), TypeError, invalid);
  }
});

test("errors carry a stable code without exposing bounded remote text by default", async () => {
  // Keep the complete JSON envelope within the client's error-body budget so
  // this assertion tests stable-code preservation rather than an impossible
  // parse of deliberately truncated JSON. Oversized non-JSON bodies are tested
  // separately through the status-code fallback contract.
  const remote = "provider-secret".repeat(1_000);
  const payload = JSON.stringify({ code: "org_taken", message: remote });
  assert.ok(new TextEncoder().encode(payload).byteLength <= MAX_ERROR_BODY_BYTES);
  const fakeFetch = async () => new Response(payload, { status: 409 });
  const client = new ZedClient({ registryUrl: "https://x.test///", fetchImpl: fakeFetch });
  await assert.rejects(
    () => client.claimOrg("acme"),
    (error) => {
      assert.ok(error instanceof ZedApiError);
      assert.equal(error.code, "org_taken");
      assert.equal(error.status, 409);
      assert.equal(error.message, "registry error 409: org_taken");
      assert.ok(!error.message.includes("provider-secret"));
      assert.ok(error.registryMessage.length <= MAX_ERROR_BODY_BYTES + 1);
      return true;
    },
  );
});

test("error code falls back to http status and non-JSON text stays explicit only", async () => {
  const fakeFetch = async () => new Response("bad gateway", { status: 502 });
  const client = new ZedClient({ registryUrl: "https://x.test", fetchImpl: fakeFetch });
  await assert.rejects(
    () => client.search("x"),
    (error) =>
      error instanceof ZedApiError &&
      error.code === "http_502" &&
      error.message === "registry error 502: http_502" &&
      error.registryMessage === "bad gateway",
  );
});

test("public reads omit bearer auth, authenticated writes attach it, and redirects are refused", async () => {
  const seen = [];
  const fakeFetch = async (url, init) => {
    seen.push({
      url: String(url),
      authorization: new Headers(init.headers).get("authorization"),
      redirect: init.redirect,
      body: init.body,
    });
    if (String(url).endsWith("/v1/orgs")) {
      return new Response(JSON.stringify({ slug: "acme", created: true }), { status: 200 });
    }
    return new Response(JSON.stringify({ query: "http", items: [] }), { status: 200 });
  };
  const client = new ZedClient({
    registryUrl: "https://x.test/gateway",
    token: " zpkg_t ",
    fetchImpl: fakeFetch,
  });
  await client.search("http");
  await client.claimOrg("acme");
  assert.equal(seen[0].url, "https://x.test/gateway/v1/search?q=http");
  assert.equal(seen[0].authorization, null);
  assert.equal(seen[0].redirect, "error");
  assert.equal(seen[1].authorization, "Bearer zpkg_t");
  assert.equal(seen[1].redirect, "error");
});

test("yank is an authenticated core operation with the canonical body", async () => {
  let captured;
  const fakeFetch = async (url, init) => {
    captured = { url: String(url), init };
    return new Response(
      JSON.stringify({ org: "acme", name: "kit", version: "1.2.0", yanked: true }),
      { status: 200 },
    );
  };
  const client = new ZedClient({ registryUrl: "https://x.test", token: "token", fetchImpl: fakeFetch });
  const result = await client.yank("acme", "kit", "1.2.0", true);
  assert.equal(result.yanked, true);
  assert.equal(captured.url, "https://x.test/v1/packages/acme/kit/versions/1.2.0/yank");
  assert.equal(captured.init.method, "POST");
  assert.equal(new Headers(captured.init.headers).get("authorization"), "Bearer token");
  assert.deepEqual(JSON.parse(captured.init.body), { yanked: true });
});

test("download omits auth, refuses redirects, and verifies sha256", async () => {
  const body = new TextEncoder().encode("artifact-bytes");
  let captured;
  const fakeFetch = async (url, init) => {
    captured = { url: String(url), init };
    return new Response(body, { status: 200 });
  };
  const client = new ZedClient({
    registryUrl: "https://x.test",
    token: "zpkg_t",
    fetchImpl: fakeFetch,
  });
  const version = makeVersion({
    sha256: sha256Hex(body),
    size: body.length,
    download_url: "https://cdn.example/artifact?signature=one",
  });
  const out = await client.downloadArtifact(version);
  assert.equal(new Headers(captured.init?.headers).get("authorization"), null);
  assert.equal(captured.init.redirect, "error");
  assert.equal(captured.url, "https://cdn.example/artifact?signature=one");
  assert.deepEqual(out, body);
});

test("download resolves registry-relative URLs and rejects insecure schemes", async () => {
  const body = new TextEncoder().encode("ok");
  let seenUrl;
  const fakeFetch = async (url) => {
    seenUrl = String(url);
    return new Response(body, { status: 200 });
  };
  const client = new ZedClient({ registryUrl: "https://x.test/gateway", fetchImpl: fakeFetch });
  await client.downloadArtifact(
    makeVersion({
      sha256: sha256Hex(body),
      size: body.length,
      download_url: "artifacts/hash",
    }),
  );
  assert.equal(seenUrl, "https://x.test/gateway/artifacts/hash");

  for (const url of ["http://evil.example/artifact", "file:///etc/passwd"]) {
    await assert.rejects(
      () => client.downloadArtifact(makeVersion({ download_url: url })),
      (error) => error instanceof ZedApiError && error.code === "insecure_download_url",
    );
  }
});

test("download allows loopback HTTP and enforces streamed and declared caps", async () => {
  const body = new TextEncoder().encode("ok");
  const client = new ZedClient({
    registryUrl: "https://x.test",
    fetchImpl: async () => new Response(body, { status: 200 }),
  });
  const out = await client.downloadArtifact(
    makeVersion({
      sha256: sha256Hex(body),
      size: body.length,
      download_url: "http://127.0.0.1:8080/artifact",
    }),
  );
  assert.deepEqual(out, body);

  const limit = 1 + 1024 * 1024;
  const big = new Uint8Array(limit + 64);
  const streamed = new ZedClient({
    registryUrl: "https://x.test",
    fetchImpl: async () => new Response(big, { status: 200 }),
  });
  await assert.rejects(
    () =>
      streamed.downloadArtifact(
        makeVersion({ sha256: sha256Hex(big), size: 1, download_url: "https://cdn.example/a" }),
      ),
    (error) => error instanceof ZedApiError && error.code === "artifact_too_large",
  );

  const declared = new ZedClient({
    registryUrl: "https://x.test",
    fetchImpl: async () =>
      new Response(body, {
        status: 200,
        headers: { "content-length": String(MAX_ARTIFACT_BYTES + 1) },
      }),
  });
  await assert.rejects(
    () =>
      declared.downloadArtifact(
        makeVersion({ sha256: sha256Hex(body), download_url: "https://cdn.example/a" }),
      ),
    (error) => error instanceof ZedApiError && error.code === "artifact_too_large",
  );
});
